pkg/bpf/bpffs_migrate.go

```diff
@@ -185,10 +187,50 @@ func finalizeMap(bpffsPath, name string, revert bool) error {
 		return nil
 	}
 
+	// Never unpin prog arrays. Doing so causes existing BPF progs to miss tail
```

Comment:
When this says "Never", what exactly does that mean? Does that mean we will leak these maps upon each upgrade? How will they get cleaned up?

Reply:
:wave: Rationale here: https://github.com/cilium/cilium/issues/16561#issuecomment-938534720

Reply:
> Otherwise we are forever consuming resources in the kernel based on the number of times we restart Cilium.

As mentioned in the linked comment: `Prog array migration only happens on Cilium up/downgrade, which is not supposed to happen all too frequently`. For example, https://github.com/cilium/cilium/pull/17758 proposes to bump `CILIUM_CALL_SIZE` from 25 to 27, which will result in that tail call map being recreated with the new size, once, when the agent is upgraded. This leaves one lingering prog array with 25 max entries on the system.

The alternative is periodically scanning all progs in the kernel (which can run into the thousands on busy nodes) for references to this old map. This process might need to go on forever until all pods using the map have been killed/rotated. Scanning takes (at least) one syscall per prog in the kernel.

Not accounting for repeated up/downgrades that will create one backup map per tick, I'm not convinced that adding a GC is worth the development and runtime execution overhead given the small size of these maps. Alternatively, we could always document that these map pins can be safely deleted if all managed pods have been rotated on a node post-upgrade.

Reply:
Is there a solution where we just delete this `backup` directory, thereby unpinning after some time? Or track `backup1` and `backup2` and then upon upgrade, we always move the current ones to `backup1` and delete `backup2`?

Reply:
> Is there a solution where we just delete this `backup` directory, thereby unpinning after some time?

Note that deleting an in-use tail call map results in dropped packets. Establishing the `after some time` heuristic is the problem here. For example, interrupted connectivity 10 mins after an upgrade would be even more surprising than interruptions during an upgrade.

> Or track backup1 and backup2 and then upon upgrade, we always move the current ones to backup1, move the old backup1 to backup2 and make sure the old backup2 are deleted?

Not sure how we would track whether `backup1` is current, old or new without keeping persistent state on the host. It's not clear what exactly you're proposing here. As Paul mentioned above, it's not even guaranteed that all pods created using Cilium version n-2 have died before the system was upgraded to n-1. There is always a risk of interruptions when blindly (without refscanning) unpinning a tail call map.

Reply:
:+1: indeed I agree that "after some fixed time" is not a reasonable solution.

One thing I don't understand is why we don't know if the map is still in use or not. In general, upon startup Cilium is intended to replace all running eBPF programs (previously installed by earlier Cilium) with new programs to ensure that the datapath is now running the new datapath implementation code. We should know exactly all of the possible points that should be configured and attached. Broadly this consists of the core programs eg attached to netdevices, sockets etc. plus per-endpoint programs. Then if there is anything else left over that we didn't migrate to the new version, we should be able to safely remove it. We already persist the information necessary to determine this.

Reply:
> We can know that, but we'd have to parse the xlated BPF programs from the kernel to search for all references to the tail call map.

Where I'm coming from is, we have a set of known intended state in the agent, and we will then conform all the hook points that we control to the intended state. This approach says "I don't want to analyse the current situation from the kernel perspective and handle all possible references that exist", it says "If there are known paths on the filesystem for pinned programs, we will move those out of the way and then instantiate the new programs in the old places. When I am done, everything should be in the correct state and there should be no remaining references to the old programs.". If we are still missing cases there, then we aren't doing enough up-front tracking of the intended state of the system.

> It is not because you're both talking minutes. Are your concerns with that approach still valid if we delete maps older than a week?

I think ultimately this is worse. Rather than "hmm, connectivity for my pod breaks after N minutes", we receive reports of "connectivity randomly broke" and then when we ask "Did you upgrade recently?" and they say "well maybe a week ago" then we will really scratch our head about whether the failure is related to upgrade or not. I'd much rather do the wrong thing early with some possibility of understanding the trigger rather than doing the wrong thing much later and have no idea why it occurred.

Reply:
> Rather than "hmm, connectivity for my pod breaks after N minutes following upgrade", we receive reports of "connectivity randomly broke" and then when we ask "Did you upgrade recently?" and they say "well maybe a week ago" then we will really scratch our head about whether the failure is related to upgrade or not.

Uh, wouldn't this only happen if the endpoints fail to be regenerated for over a week? I think we have more serious issues if that happens :grimacing:

Reply:
Is that true? That feels like it frames the resource cleanup question much more nicely.

If everything succeeds regeneration, delete the `backup` directory. Otherwise we need to keep it around to keep the endpoints successfully connected, and hopefully the warning signs from regeneration will reach the user and they'll roll back to a known good state. In that particular case there would still be a `backup` directory on the system until they next upgrade, but at that point we could just remove/overwrite those `backup` pins and hopefully have better luck the second time.